File: solutions/0298-binary-tree-longest-consecutive-sequence/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int longestConsecutive(TreeNode root) {
        int[] result = new int[1];

        //Edge case 
        if(root == null) {
            return 0;
        }

        dfs(root,root.val,-1,result);

        return result[0];
        
    }

    private void dfs(
        TreeNode node,
        int parentValue,
        int lengthSoFar,
        int[] result
    ) {

        // we need to know the parent value to check to make sure we have the longest path

        if(node.val == parentValue + 1) {
            lengthSoFar = lengthSoFar + 1;
        }else {
            lengthSoFar = 1;
        }

        //update the global max 
        if(lengthSoFar > result[0]) {
            result[0] = lengthSoFar;
        }

        //Base case 

        if(node.left == null && node.right == null) {
            //Do nothing
            ;
        }

        //Recursive case 
        if(node.left != null) {
            dfs(node.left,node.val,lengthSoFar,result);
        }

        if(node.right != null) {
            dfs(node.right,node.val,lengthSoFar,result);
        }
    }
}

File: solutions/0298-binary-tree-longest-consecutive-sequence/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestConsecutive(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        globalAns = [0]

        def dfs(node,pval,lengthSofar):

            if(node.val == pval+1):
                lengthSofar += 1
            else:
                lengthSofar = 1

            if lengthSofar > globalAns[0]:
                globalAns[0] = lengthSofar
            
            #Recursive Cas
            if node.left is not None:
                dfs(node.left,node.val,lengthSofar)
            if node.right is not None:
                dfs(node.right,node.val,lengthSofar)

        dfs(root,root.val,1)        
        return globalAns[0]

               
        
